src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/UniversalCryptoOneShot.cs

```diff
@@ -20,31 +20,137 @@ internal static class UniversalCryptoOneShot
+            // If no padding is going to removed, then we already know the buffer is too small
+            // since that requires a buffer at-least the size of the ciphertext. Bail out early.
+            // The second condition is where the output length is short by more than a whole block.
+            // All valid padding is at most one complete block. If the difference between the
+            // output and the input is more than a whole block then we know the output is too small.
+            if (!SymmetricPadding.DepaddingRequired(paddingMode) ||
+                input.Length - cipher.BlockSizeInBytes > output.Length)
+            {
                 bytesWritten = 0;
                 return false;
             }
 
+            // At this point the destination might be big enough but we don't know until we've
+            // transformed the last block, and input is within one block of being the right
+            // size.
+            // For sufficiently small ciphertexts, do them on the stack.
+            const int MaxInStackDecryptionBuffer = 128;
+            Span<byte> stackBuffer = stackalloc byte[MaxInStackDecryptionBuffer];
+
+            if (input.Length <= MaxInStackDecryptionBuffer)
+            {
+                int stackTransformFinal = cipher.TransformFinal(input, stackBuffer);
+                int depaddedLength = SymmetricPadding.GetPaddingLength(
+                    stackBuffer.Slice(0, stackTransformFinal),
+                    paddingMode,
+                    cipher.BlockSizeInBytes);
+
+                if (output.Length < depaddedLength)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+
+                stackBuffer.Slice(0, depaddedLength).CopyTo(output);
+                bytesWritten = depaddedLength;
+                return true;
+            }
+
+            // If the source and destination do not overlap, we can decrypt directly in to the user buffer.
+            if (!input.Overlaps(output))
+            {
+                // At this point we know that we have multiple blocks that need to be decrypted.
+                // So we decrypt all but the last block directly in to the buffer. The final
+                // block we decrypt in to a stack buffer, and if it fits, copy the last block to
+                // the output.
+
+                // We should only get here if we have multiple blocks to transform. The single
+                // block case should have happened on the stack.
+                Debug.Assert(input.Length > cipher.BlockSizeInBytes);
+
+                int writtenToOutput = 0;
+
+                // CFB8 means this may not be an exact multiple of the block size.
+                // If the an AES CFB8 ciphertext length is 129 with PKCS7 padding, then
+                // we'll have 113 bytes in the unpaddedBlocks and 16 in the paddedBlock.
```

Comment:
I assume we have a test for this specific set of weirdness? :smile:

Reply:
Should; but I lumped this in with my "Better unit tests on boundary conditions" TODO.

Reply:
This is covered by the tests added in #64949, so I would say we are good here.